frontend/internal-packages/agent/src/chat/workflow/shared/langGraphUtils.ts

```diff
@@ -4,40 +4,87 @@ import type { Usecase } from '../../../langchain/agents/qaGenerateUsecaseAgent/a
 import type { Repositories } from '../../../repositories'
 import type { NodeLogger } from '../../../utils/nodeLogger'
 
-export const DEFAULT_RECURSION_LIMIT = 10
+/**
+ * Default recursion limit for LangGraph workflow execution.
+ * This value limits the total number of state transitions (edges) in the graph.
+ *
+ * The workflow has 9 nodes with a maximum of 3 retries per node:
+ * - Normal execution: 10 transitions (START → 9 nodes → END)
+ * - With retries: up to 37 transitions (10 normal + 27 retry transitions)
+ *
+ * Setting this to 40 ensures:
+ * - Complete workflow execution under normal conditions
+ * - Sufficient headroom for error handling and retries
+ * - Protection against infinite loops
+ */
+export const DEFAULT_RECURSION_LIMIT = 40
 
 /**
  * Create LangGraph-compatible annotations (shared)
  */
 export const createAnnotations = () => {
   return Annotation.Root({
-    userInput: Annotation<string>,
+    userInput: Annotation<string>({
+      reducer: (_, newValue: string) => newValue,
+      default: () => '',
+    }),
```

Comment:
Here's what happened:

LangGraph's default `LastValue` annotation only allows one write per channel per workflow step. However, when our nodes retry due to errors, they spread the entire state (...state), which includes all existing channel values.

This caused LangGraph to think we were writing to channels like userInput multiple times in the same step, triggering `"Invalid update for channel"` errors.

**Adding custom reducers** that explicitly handle duplicate values (by simply returning the new value) tells LangGraph how to resolve these "conflicts" and allows retries to work properly.

about reducer
https://langchain-ai.github.io/langgraph/concepts/low_level/#default-reducer